File: src/main/java/core/procedures/strings/MakeString.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMClass;
import core.scm.SCMMutableString;

public final class MakeString extends AFn {

  public MakeString() {
    super(new FnArgsBuilder().min(1).max(2)
                             .mandatory(new Class[]{SCMClass.ExactNonNegativeInteger.class})
                             .rest(Character.class).build());
  }

  @Override
  public String getName() {
    return "make-string";
  }

  @Override
  public SCMMutableString apply(Object... args) {
    Long s = ((Number)args[0]).longValue();
    Object c = (args.length == 1) ? Character.MIN_VALUE : args[1];
    SCMMutableString string = new SCMMutableString();
    for (long i = 0; i < s; i++) {
      string.append(c);
    }
    return string;
  }
}

File: src/main/java/core/procedures/strings/MakeString.kt
package core.procedures.strings

import core.procedures.AFn
import core.procedures.Arity.Range
import core.scm.MutableString
import core.scm.Type

class MakeString : AFn<Any?, MutableString>(name = "make-string", isPure = true, arity = Range(1, 2),
                                            mandatoryArgsTypes = arrayOf(Type.ExactNonNegativeInteger::class.java),
                                            restArgsType = Char::class.javaObjectType) {

    override operator fun invoke(args: Array<out Any?>): MutableString {
        val length = (args[0] as Number).toInt()
        val c = if (args.size == 1) Character.MIN_VALUE else args[1]
        return MutableString(length).apply { for (i in 0 until length) { append(c) } }
    }
}
